### PR TITLE
Apply block-level bucketing

### PR DESCRIPTION
**Pull request by ruisizhang123**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #19
* #18
* #17
* #16

### CODE REVIEW

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -1031,6 +1031,11 @@ class halide:
     scan_kernels = False
 
 
+class simplefsdp:
+    enable_bucket = True
```
Comment (tianyu-l) [inline torch/_inductor/config.py]: let's default to False, so for non-Transformer structure it's safe to use, even without setting this flag

File: torch/_inductor/simple_fsdp/transformer_block_bucket.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+    for idx, node in enumerate(snodes):
+        current_module = node_block_list[idx]
+        if current_module != last_module and len(all_gather_list) > 0:
+            # bucketing in the block boundary
+            assert len(all_gather_list) == len(ag_wait_list)
+            merged_all_gather = merge_allgather(sched, all_gather_list, all_gather_dep_list)
+            merged_wait = merge_ag_wait(sched, ag_wait_list, all_gather_list, merged_all_gather[-1].node)
+    
+            for n in  merged_all_gather+merged_wait:
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/transformer_block_bucket.py]: linting?

**Review by tianyu-l — COMMENTED** [review]
The code looks organized, except for the hacky `merge_block_name` method. That's understandable as there might be bugs in annotation from upstream, e.g. aot autograd or graph lowering. We should create GH issues in pytorch if that's the case.

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -1596,17 +1596,33 @@ def __init__(self, nodes: List[ir.Operation]) -> None:
...
+                self, self.nodes
             )
-        elif self.post_grad_graph_id == 1:
-            # reorder backward graph
-            self.nodes = reorder.reorder_all_gather(
-                self.nodes, all_gather_before_last_wait=False
-            )
-            self.nodes = reorder.reorder_reduce_scatter(self.nodes)
+
+            if self.post_grad_graph_id == 1:
```
Comment (tianyu-l) [inline torch/_inductor/scheduler.py:1609]: Let's figure out what's the best way to tell if a graph is forward/backward. We can ask Will on this. Maybe add a TODO item here if you can't figure out it very quickly.

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -1031,6 +1031,11 @@ class halide:
     scan_kernels = False
 
 
+class simplefsdp:
```
Comment (tianyu-l) [inline torch/_inductor/config.py]: later on we can have another config, e.g. `bucket_strategy` and choose from `transformer_block` vs. `greedy` vs. `dynamic_programming`, etc.

**Review by tianyu-l — APPROVED** [review]

File: torch/_inductor/scheduler.py (hunk 3 above)
Comment (ruisizhang123) [reply torch/_inductor/scheduler.py:1609]: added

File: torch/_inductor/config.py (hunk 4 above)
Comment (ruisizhang123) [reply torch/_inductor/config.py]: sure, sg

**Pull request merged**